### PR TITLE
Fix multiple correctness issues

### PR DESCRIPTION
**Pull request by gwynne**
Here's a list of changes:

- `StackTrace` no longer severely misuses pointers or invokes undefined behavior when gathering backtraces on Linux. It also no longer has a number of implicit fatal error code paths.
- Several deprecation warnings that show up when building with 5.8+ are gone.
- Concurrency support now correctly back-deploys as far as macOS 10.15. As a side effect, the minimum Swift version has increased from 5.5 to 5.5.2.
- The `Deployment` example target has been significantly improved and no longer references local paths on @tanner0101's computer 😆
- The TOTP/HOTP logic is now both faster and safer (no more use of unsafe pointers, in particular).
- `DecoderWrapper`, an unsafe `Codable` hack with serious pitfalls, has been hard-deprecated.
- `PlaintextEncoder` and `PlaintextDecoder` have been significantly cleaned up and improved. All `fatalError()`s have been removed, performance has been noticeably improved, and errors are more accurate.
- `URLEncodedFormEncoder` and `URLEncodedFormDecoder` got the same treatment.
- The entire `Validation` submodule has been overhauled to fix systemic misuses of `Codable` that were starting to cause noticeable problems.
- As a side effect of said overhaul, `ValidationKey` is now deprecated in favor of using the essentially identical `BasicCodingKey` type.
- `ContentContainer` and `URLQueryContainer` received the same "fix `Codable` usage" treatment.
- `ContentConfiguration` now throws more useful errors.
- When `CodingKeyRepresentable` is available in the stdlib (Swift 5.6+), Vapor's protocol of the same name now becomes a `typealias` for it.

### CODE REVIEW

**Review by 0xTim — APPROVED** [review]
Couple of questions but LGTM!

File: Sources/Vapor/Content/JSONCoders+Content.swift [hunk 1]
```diff
@@ -5,17 +5,58 @@ import NIOHTTP1
...
+    }
+    
+    public func encode<E>(_ encodable: E, to body: inout ByteBuffer, headers: inout HTTPHeaders, userInfo: [CodingUserInfoKey: Any]) throws
+        where E: Encodable
     {
         headers.contentType = .json
-        try body.writeBytes(self.encode(encodable))
+        
+        if !userInfo.isEmpty {
+            let actualEncoder = JSONEncoder()
```
[the remark this answers is not in the document]
Comment (gwynne) [reply Sources/Vapor/Content/JSONCoders+Content.swift:18]: Bingo. It's thread safe to actually invoke the encoder from multiple threads, but changing the user info is a mutation, so we need a copy for that case.

File: Sources/Vapor/Content/PlaintextDecoder.swift [hunk 2]
```diff
@@ -3,95 +3,80 @@ import NIOHTTP1
...
+        guard let result = L.init(value) else { throw DecodingError.dataCorruptedError(in: self, debugDescription: "Could not decode \(L.self) from \"\(value)\"") }
+        return result
     }
 
-    func decodeNil() -> Bool {
-        if let plaintext = decoder.plaintext {
-            return plaintext.isEmpty
+    func decode(_ type: String.Type) throws -> String { self.plaintext ?? "" }
+
+    func decode(_: Bool.Type) throws -> Bool { try self.losslessDecode(Bool.self) }
```
[the remark this answers is not in the document]
Comment (gwynne) [reply Sources/Vapor/Content/PlaintextDecoder.swift:57]: Well that's the trick, isn't it? Implementing the non-generic dispatch methods on Codable containers can yield a shockingly large speedup.

File: Sources/Vapor/Utilities/BasicCodingKey.swift [hunk 3]
```diff
@@ -1,31 +1,28 @@
-/// Capable of being represented by a `CodingKey`.
+#if swift(>=5.6) && os(Linux)
```
[the remark this answers is not in the document]
Comment (gwynne) [reply Sources/Vapor/Utilities/BasicCodingKey.swift:1]: There's also the platform condition involved, so there'd still be a conditional here; dropping 5.5 wouldn't really gain anything.

File: Sources/Vapor/Content/JSONCoders+Content.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xTim) [reply Sources/Vapor/Content/JSONCoders+Content.swift:18]: Want to add a comment to make that clear for future users (and I mean me) so we don't revert it 😅

File: Sources/Vapor/Content/PlaintextDecoder.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (0xTim) [reply Sources/Vapor/Content/PlaintextDecoder.swift:57]: Again might be worth a comment

File: Sources/Vapor/Content/ContentConfiguration.swift [hunk 4]
```diff
@@ -104,11 +102,9 @@ public struct ContentConfiguration {
     ///
     ///     let coder = try coders.requireHTTPEncoder(for: .json)
     ///
-    /// - parameters:
-    ///     - HTTPMediaType: An encoder for this `MediaType` will be returned.
```
Comment (0xTim) [inline Sources/Vapor/Content/ContentConfiguration.swift:108]: The lines above this for this functions comments are out of date and need DocCing

File: Sources/Vapor/Content/ContentConfiguration.swift [hunk 5]
```diff
@@ -117,44 +113,31 @@ public struct ContentConfiguration {
     /// Returns a `HTTPMessageDecoder` for the specified `MediaType` or throws an error.
     ///
     ///     let coder = try coders.requireHTTPDecoder(for: .json)
-    ///
```
Comment (0xTim) [inline Sources/Vapor/Content/ContentConfiguration.swift:120]: Again comment block out of date

File: Sources/Vapor/URLEncodedForm/URLEncodedFormDecoder.swift [hunk 6]
```diff
@@ -105,9 +119,15 @@ public struct URLEncodedFormDecoder: ContentDecoder, URLQueryDecoder {
     ///     - configuration: Overrides the default coding configuration
     /// - returns: An instance of the `Decodable` type (`D`).
     /// - throws: Any error that may occur while attempting to decode the specified type.
-    public func decode<D>(_ decodable: D.Type, from string: String) throws -> D where D : Decodable {
+    public func decode<D>(_ decodable: D.Type, from string: String, userInfo: [CodingUserInfoKey: Any] = [:]) throws -> D where D : Decodable {
```
Comment (0xTim) [inline Sources/Vapor/URLEncodedForm/URLEncodedFormDecoder.swift:122]: The doc comments for this function need updating

File: Sources/Vapor/URLEncodedForm/URLEncodedFormEncoder.swift [hunk 7]
```diff
@@ -72,19 +75,32 @@ public struct URLEncodedFormEncoder: ContentEncoder, URLQueryEncoder {
...
-    /// `URLContentEncoder` conformance.
+    /// ``URLQueryEncoder`` conformance.
     public func encode<E>(_ encodable: E, to url: inout URI) throws
         where E: Encodable
     {
-        url.query = try self.encode(encodable)
+        try self.encode(encodable, to: &url, userInfo: [:])
+    }
+    
+    public func encode<E>(_ encodable: E, to url: inout URI, userInfo: [CodingUserInfoKey: Any]) throws
```
Comment (0xTim) [inline Sources/Vapor/URLEncodedForm/URLEncodedFormEncoder.swift:100]: Missing DocC comments

File: Sources/Vapor/URLEncodedForm/URLEncodedFormEncoder.swift [hunk 8]
```diff
@@ -98,10 +114,14 @@ public struct URLEncodedFormEncoder: ContentEncoder, URLQueryEncoder {
     ///     - configuration: Overrides the  coding config for this encoding call.
     /// - returns: Encoded `Data`
     /// - throws: Any error that may occur while attempting to encode the specified type.
-    public func encode<E>(_ encodable: E) throws -> String
+    public func encode<E>(_ encodable: E, userInfo: [CodingUserInfoKey: Any] = [:]) throws -> String
```
Comment (0xTim) [inline Sources/Vapor/URLEncodedForm/URLEncodedFormEncoder.swift:117]: Function comments are out of date

**Review by 0xTim — APPROVED** [review]
Looks good. I've spotted a few places where the comments are out of date and noted them, they should probably be fixed before merging

**Pull request merged**